chainer/functions/math/matmul.py

```diff
@@ -36,36 +34,20 @@ def _get_ld(a):
 
 
 def _matmul(a, b, transa=False, transb=False, transout=False):
-    a = array.as_mat(a)
-    b = array.as_mat(b)
     if transout:
-        # (A B)^T = B^T A^T
         transa, transb = not transb, not transa
         a, b = b, a
-    if transa:
-        a = a.T
-    if transb:
-        b = b.T
-    return a.dot(b)
-
-
-def _batch_matmul(a, b, transa=False, transb=False, transout=False):
-    a = a.reshape(a.shape[:2] + (-1,))
-    b = b.reshape(b.shape[:2] + (-1,))
-    trans_axis = (0, 2, 1)
-    if transout:
-        transa, transb = not transb, not transa
-        a, b = b, a
-    if transa:
-        a = a.transpose(trans_axis)
-    if transb:
-        b = b.transpose(trans_axis)
+    if transa and a.ndim != 1:
+        a = a.swapaxes(-1, -2)
+    if transb and b.ndim != 1:
+        b = b.swapaxes(-1, -2)
     xp = cuda.get_array_module(a)
+
     if xp is numpy:
-        ret = numpy.empty(a.shape[:2] + b.shape[2:], dtype=a.dtype)
-        for i in six.moves.range(len(a)):
-            ret[i] = numpy.dot(a[i], b[i])
-        return ret
+        if a.ndim <= 2:
+            return numpy.dot(a, b)
+        else:
+            return numpy.einsum('...ij,...jk->...ik', a, b)
     return xp.matmul(a, b)
```

Comment:
`if xp is numpy and numpy.version.version.split('.')[1] == '9'` is prefer in L.46?